src/utils/encoding.ts

```diff
@@ -0,0 +1,56 @@
+import fs from "fs";
+import iconv from "iconv-lite";
+import readline from "readline";
+import { extractFileFormat } from "./file";
+import type { SupportedImportEncoding } from "./file";
+
+export const isMismatchEncoding = async (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+): Promise<boolean> => {
+  const format = extractFileFormat(filePath);
+  let isMismatch = false;
+  switch (format) {
+    case "csv":
+      isMismatch = await isMismatchEncodingOfCsvFile(filePath, encoding);
+      break;
+  }
+
+  return isMismatch;
+};
+
+const isMismatchEncodingOfCsvFile: (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+) => Promise<boolean> = async (filePath, encoding) => {
+  const decodedFirstLine = await getDecodedFirstLine(filePath, encoding);
+  return isContainUntranslatableChars(decodedFirstLine);
+};
+
+const getDecodedFirstLine: (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+) => Promise<string> = async (filePath, encoding) => {
+  const stream = fs.createReadStream(filePath);
+  const readWriteStream = stream.pipe(iconv.decodeStream(encoding));
```

Comment:
This might cause an error if input CSV includes different encoding text in their data,
because stream drains data automatically, not only the first line.

Reply:
@mshrtsr 

Ohmmm.
As I understand, the readable.pipe() method is used to connect the output of a readable stream to the input of a writable stream. And, this method is commonly used to efficiently transfer data from one stream to another without loading the entire data into memory.
So, I wonder why do we need to worry about the remaining lines if we only check the first line of the CSV file?

In addition, if `iconv-lite` cannot decode the first line, It just replaces by untranslatable characters. So, how can the error happen?

Reply:
@hung-cybo 
I see.
According to the following reasons, let's keep this code as is.
- iconv-lite doesn't throw an error and uses untranslatable chars when decoding fails
- We cannot use `iconv.decode` function for each line generated from `readline`
  - https://github.com/kintone/cli-kintone/pull/437#discussion_r1296987031

---
Short description of why we need to consider the remaining lines.

The stream drains chunks until buffer usage reaches the `highWaterMark`.
Even if only the first line needs to be decoded, the second and later lines are actually decoded, too.
I created the sample code.

```typescript
const fsStream = fs.createReadStream("input_sjis.csv");
// logger prints and passes through chunks
const logger = new Transform({
  transform(chunk, encoding, callback) {
    console.log("chunk received");
    console.log(chunk + "");
    callback(null, chunk);
  },
});

console.log("pipe to decoder and logger");
const decodedStream = fsStream
  .pipe(iconv.decodeStream("sjis"))
  .pipe(logger);
console.log("pipe to readline");
const readlineInterface = readline.createInterface({
  input: decodedStream,
});
console.log("read the first row");
const { value: firstRow } = await readlineInterface[
  Symbol.asyncIterator
]().next();

console.log(firstRow);
fsStream.close();
```

The result is the following.
As you can see, chunks are moved through the decoder and logger, and arrived to `readlineInterface` even if we read only one line from the `readlineInterface`.
If an error occurs after the second line in the stream, an unhandled error will happen.

```log
  console.log
    pipe to decoder and logger

      at Object.log (src/utils/__tests__/encoding.test.ts:33:13)

  console.log
    pipe to readline

      at Object.log (src/utils/__tests__/encoding.test.ts:37:13)

  console.log
    read the first row

      at Object.log (src/utils/__tests__/encoding.test.ts:41:13)

  console.log
    chunk received

      at Transform.log [as _transform] (src/utils/__tests__/encoding.test.ts:27:17)

  console.log
    "レコード番号","更新者","作成者","文字列__1行_","ラジオボタン","更新日時","作成日時"
    "1","tasshi","tasshi","文字列１２３","選択肢1","2022-05-02T07:58:00Z","2022-05-02T07:58:00Z"

      at Transform.log [as _transform] (src/utils/__tests__/encoding.test.ts:28:17)

  console.log
    "レコード番号","更新者","作成者","文字列__1行_","ラジオボタン","更新日時","作成日時"

      at Object.log (src/utils/__tests__/encoding.test.ts:46:13)
```

Reply:
@mshrtsr 

Oh, now I got it.
Thank you for your explanation.